### CS/ECE/ISyE 524 &mdash; Introduction to Optimization &mdash; Summer 2021 ###

# Optimizing COVID-19 Vaccine Delivery #

#### Abishek Kumar (akumar225@wisc.edu) and Ranganath (Bujji) Selagamsetty (selagamsetty@wisc.edu)

### Table of Contents

1. [Introduction](#1.-Introduction)
1. [Mathematical Model](#2.-Mathematical-model)
1. [Solution](#3.-Solution)
1. [Results and Discussion](#4.-Results-and-discussion)
1. [Conclusion](#5.-Conclusion)

## 1. Introduction ##

### Overview ###

The novel coronavirus virus (COVID - 19) began its worldwide spread on [January 9th, 2020](https://www.ajmc.com/view/a-timeline-of-covid19-developments-in-2020), going on to infect hundreds of thousands of people. It has been almost 19 months since the virus first broke, and one thing that has been clear to see is the lack of sophistication in the ability of the United States to produce, store, and distribute COVID-19 vaccines. 

Of the many large pharmaceutical companies in the US, Pfizer, Moderna, and Johnson & Johnson were the first three to commercially produce and administer these vaccines. While having a more than one vaccine provider was advantageous, it posed logistical challenges due to the manufacturing differences, and gave rise to an interesting optimization question: What is the most cost effective way to produce a COVID-19 vaccines and distribute them across the nation?

To increase the modelling fidelity and decrease the representational gap, we attempted to design the model based off of real world information (e.g., actual locations that source ingredients for vaccine production, locations that manufacture the vaccine, and actual locations that would need the vaccine) . However, for the sake of seeing more interesting results, we decided to use artificial values for many of the modelling data elements (this is further explained in the following sections).

### Sources Map  ###

Below, we have a map of what we call the "sources" in our model. These are the locations that at which ingredients for vaccine production can be sourced from. Since Pfizer is the [largest](https://www.imarcgroup.com/top-players-global-active-pharmaceutical-ingredients-market) [Active Pharmaceutical Ingredient (API)](https://www.contractpharma.com/contents/view_online-exclusives/2020-03-05/api-supply-for-us-market/) company in the US, we chose our ingredient sourcing sites from the list of [Pfizer ingredient manufacturing locations](https://www.pfizer.com/products/pfizer-global-supply/us-manufacturing-sites/): Andover, MA; McPherson, KA; Sanford, NC.

<img src="files/imgs/src_nodes.png">

### Sites Map ###

Below, we have a map of what we call the "sites" in our model. These are the locations that at which ingredients for vaccine production are collected and the vaccines are actually manufactued. Pfizer has two vaccine production centers in the US, but for this study we will only look at the larger one: [Kalamazoo, Michigan](https://www.pfizer.com/science/coronavirus/vaccine/manufacturing-and-distribution). Moderna relied on a third party drug manufacturer, Catalent, to produce it's vaccine. Catalent had a single production site in the US at [Bloomington, Indiana](https://www.catalent.com/catalent-news/moderna-and-catalent-announce-long-term-strategic-collaboration-for-dedicated-vial-filling-of-modernas-covid-19-vaccine-and-clinical-portfolio/#:~:text=(NYSE%3A%20CTLT)%2C%20the,vial%20filling%20line%20for%20the). Johnson and Johnson was granted oversight of the vaccine distribution center in [Baltimore, Maryland](https://www.npr.org/2021/04/04/984274691/johnson-johnson-to-oversee-vaccine-production-at-baltimore-facility), and used that location as it's base of vaccine distribution in the US. 

While we use this information as the basis for creating the topology of our network, our model allows the production of any type of vaccine (Pfizer, Moderna, or Johnson) at any of these sites. This was intentionally done to better reflect that during an emergency session, every production site would be working to produce as many of any type of vaccine when needed. 

<img src="files/imgs/prod_nodes.png">

### Destinations Map ### 

Below, we have a map of what we call the "destinations" in our model. These are the locations that vaccines need to be shipped to. These were chosen to be the three most populated cities in the US: [New York City, Los Angeles, and Chicago](https://www.investopedia.com/articles/personal-finance/050815/top-10-most-developed-cities-us.asp). The distribution of the populations of the age groups in these cities was synthetically generated. 

<img src="files/imgs/dest_nodes.png">

#### Arc Map 

Putting all the above together, below we have a map of how all the nodes in the network connect with each other. We simplify our model and allow all sources to connect with all sites and all sites to connect with all destinations, for all types of transportation methods. 

<img src="files/imgs/arcs.png">

### Project Layout ###

2. [Mathematical Model](#2.-Mathematical-model)
    
    2.1. The Word Problem - Overview of the optimization problem described in plain text
        
        2.1.a. Ingredients - The data associated with the ingredients 
        
        2.1.b. Vaccines - The data associated with the vaccines 
        
        2.1.c. Employees - The data associated with the employees 
    
    2.2 The Formal Model - Mathematical representation of the optimization problem

    2.3 Standard Form - The standard form of the optimization problem
    
    2.4 Modelling Details - Discussion and rationale of the model type 
    
3. [Solution](#3.-Solution)

4. [Results and Discussion](#4.-Results-and-discussion)

    4.1. Ingredients - Results of how the ingredients were shipped and stored
    
    4.2. Vaccines - Results of how the vaccines were shipped and stored
    
    4.3. Employees - Results of how the employees were hired and fired
    
    4.4. Limitations - Limitations of our modelling assumptions
    
5. [Conclusion](#5.-Conclusion)

    5.1. Summary
    
    5.2. Future Work
    

## 2. Mathematical model ##

### 2.1. The Word Problem ###

#### 2.1.a. Ingredients ####

As discussed in the [Introduction](#1.-Introduction) section, there are three locations to get the ingredients to produce a COVID-19 vaccine: Andover, McPherson, and Sanford. There are three possible ingredients that can used to manufacture the three types of vaccines (Pfizer, Moderna, and Johnson): ing1, ing2, and ing3. The distributions of of each ingredient needed to make each type of vaccine is shown below (Note, the number of ingredients and the distributions were synthetically generated):

|Vaccine Type| Amount (mL) of ing1 to make 1 dose of vaccine | Amount (mL) of ing2 to make 2 dose of vaccine | Amount (mL) of ing3 to make 1 dose of vaccine| 
| --- | --- | --- | --- |
| Pfizer | 16 | 20 | 9 |
| Moderna | 4 | 25 | 12 |
| Johnson | 7 | 14 | 35 |

Each sourcing location has a fixed supply of ingredients per month (values are in **thousands** and were synthetically generated) :

| Source | ing1(mL)<sub>Month=1</sub> | ing2(mL)<sub>Month=1</sub>  | ing3(mL)<sub>Month=1</sub>  | ing1(mL)<sub>Month=2</sub>  | ing2(mL)<sub>Month=2</sub> | ing3(mL)<sub>Month=2</sub> | ing1(mL)<sub>Month=3</sub> | ing2(mL)<sub>Month=3</sub> | ing3(mL)<sub>Month=3</sub> |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| Andover   | 850 | 622 | 905 | 450.4 | 142 | 405 | 140 | 182 | 205 |
| McPherson | 700 | 615 | 900 | 200.5 | 241 | 0.8 | 0.2 | 321 | 0.8 |
| Sanford   | 950 | 820 | 784 | 400.6 | 608 | 700 | 409 | 201 | 1.0045 | 

The distances between source locations and site locations is shown in the map below (values are in miles and were synthetically generated):

<img src="files/imgs/dist1.png">

Once shipped to a site location, one of two things may happen. In the first case, the ingredients are combined in the appropriate proportions to manufacture various amounts of the different types of vaccines and then shipped out to the destinations that need them. In the second case, vaccines are manufactured, but may be stored at a production site to ship out at a later date (i.e., these doses of vaccine are kept in inventory). In the latter case, there is a cost to store each ingredient (values are in dollars per mL of ingredient and were synthetically generated):

| Sites | ing1 (mL) | ing2 (mL) | ing3 (mL)  |
| --- | --- | --- | --- | 
| Kalamazoo   | \$120 | \$200| \$210|
| Bloomington | \$20  | \$35 | \$15|
| Baltimore   | \$124 | \$300| \$150|

Each method of transport packs a number of ingredients into a single shipment. These shipment sizes are shown below (values are mL of ingredient and were synthetically generated): 

| Ingredient | Air| Road | Train | 
| --- | --- | --- | --- |
| ing1 | 1200 | 5000 | 8200 |
| ing2 | 4200 | 6240 | 3540 |
| ing3 | 4710 | 5620 | 384 |

It costs different amounts to ship the various ingredients to the site locations. These values are shown below (values are dollars per shipment and were synthetically generated):

| Source| Kalamzoo<sub>Air</sub> | Kalamzoo<sub>Road</sub> | Kalamzoo<sub>Train</sub> | Bloomington<sub>Air</sub>  | Bloomington<sub>Road</sub> | Bloomington<sub>Train</sub> | Baltimore<sub>Air</sub> | Baltimore<sub>Road</sub> | Baltimore<sub>Train</sub> |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| Andover   | \$2 | \$5 | \$7  | \$1 | \$5 | \$6 | \$1 | \$4 | \$8 |
| McPherson | \$4 | \$8 | \$12 | \$2 | \$4 | \$9 | \$3 | \$6 | \$7 |
| Sanford   | \$3 | \$7 | \$8  | \$3 | \$5 | \$7 | \$2 | \$5 | \$7 |

Finally, the supply chains for each shipping method have limits, so each shipping method can only ship so many shipments of ingredients based on the month (values are in number of shipments and were synthetically generated):

| Ingredient | Air<sub>Month=1</sub> | Road<sub>Month=1</sub>  | Train<sub>Month=1</sub>  | Air<sub>Month=2</sub> | Road<sub>Month=2</sub> | Train<sub>Month=2</sub> | Air<sub>Month=3</sub> | Road<sub>Month=3</sub> | Train<sub>Month=3</sub> |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| ing1 | 30 | 90 | 200 | 15 | 45 | 800 | 65 | 95 | 100 |
| ing2 | 20 | 100 | 350 | 60 | 50 | 450 | 40 | 50 | 150 |
| ing3 | 25 | 30 | 80 | 80 | 100 | 600 | 90 | 300 | 800 |


#### 2.1.b. Vaccines ####
As discussed in the [Introduction](#1.-Introduction) section, there are three locations  at which vaccines are produced: Kalamazoo, Bloomington, and Baltimore. 

to get the ingredients to produce a COVID-19 vaccine: Andover, McPherson, and Sanford. There are three possible ingredients that can used to manufacture the three types of vaccines (Pfizer, Moderna, and Johnson): ing1, ing2, and ing3. The distributions of of each ingredient needed to make each type of vaccine is shown below (Note, the number of ingredients and the distributions were synthetically generated):

The distances between site locations and destination locations is shown in the map below (values are in miles and were synthetically generated):

<img src="files/imgs/dist2.png">

At a site location, after the ingredient have been combined to produce certain amounts of the various types of vaccines, the vaccines themselves are either shipped out to destination nodes or stored in the production site themselves (i.e., kept as inventory to ship out at a later date). In the latter case, there is a cost to store each vaccine (values are in dollars per mL of vaccine and were synthetically generated):

| Sites | Pfizer (mL) | Moderna (mL) | Johnson (mL)  |
| --- | --- | --- | --- | 
| Kalamazoo   | \$400 | \$280| \$610|
| Bloomington | \$250 | \$325| \$165|
| Baltimore   | \$380 | \$459| \$357|

Each method of transport packs a number of vaccines into a single shipment. These shipment sizes are shown below (values are mL of vaccine and were synthetically generated): 

| Vaccine | Air| Road | Train | 
| --- | --- | --- | --- |
| Pfizer | 1500 | 5000 | 840 |
| Moderna | 2000 | 6000 | 3580 |
| Johnson | 4200 | 5600 | 3400 |

It costs different amounts to ship the various vaccines to the destination locations. These values are shown below (values are dollars per shipment and were synthetically generated):

| Site| New York City<sub>Air</sub> | New York City<sub>Road</sub> | New York City<sub>Train</sub> | Los Angeles<sub>Air</sub>  | Los Angeles<sub>Road</sub> | Los Angeles<sub>Train</sub> | Chicago<sub>Air</sub> | Chicago<sub>Road</sub> | Chicago<sub>Train</sub> |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| Kalamazoo   | \$2 | \$5 | \$7  | \$1 | \$5 | \$6 | \$1 | \$4 | \$8 |
| Bloomington | \$4 | \$8 | \$12 | \$2 | \$4 | \$9 | \$3 | \$6 | \$7 |
| Baltimore   | \$3 | \$7 | \$8  | \$3 | \$5 | \$7 | \$2 | \$5 | \$7 |

The supply chains for each shipping method have limits, so each shipping can only ship so many shipments of vaccines based on the month (values are in number of shipments and were synthetically generated):

| Vaccine | Air<sub>Month=1</sub> | Road<sub>Month=1</sub>  | Train<sub>Month=1</sub>  | Air<sub>Month=2</sub> | Road<sub>Month=2</sub> | Train<sub>Month=2</sub> | Air<sub>Month=3</sub> | Road<sub>Month=3</sub> | Train<sub>Month=3</sub> |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| Pfizer  | 120 | 240 | 370 | 101 | 280 | 340 | 128 | 244 | 373 |
| Moderna | 250 | 100 | 350 | 450| 130| 300| 240| 120| 340|
| Johnson | 215 |130| 840| 245| 430| 440| 255| 330| 240 |

The demand for a vaccine was calulated in a way that is reflected of how vaaccine demands were considered in the real world; based on population sizes and age distributions. Thus, the following population data was used (values were synthetically generated):

| Destinations | Population | Young<sub>%</sub>  | Middle Aged<sub>%</sub>  | Old<sub>%</sub> |
| --- | --- | --- | --- | --- | 
| New York City | 1500 | 25 | 55 | 20 |
| Los Angeles   | 3500 | 10 | 60 | 30 |
| Chicago       | 7500 | 20 | 40 | 40 |

The line of reasoning for this was assuming that the nation would require certain percentages of the differing age groups to get doses of certain vaccines in certain months. The requirements have the following distridution (values are percentages of populations and were synthetically generated): 

| Age Group | (%) Pfizer<sub>Month=1</sub> | (%) Moderna<sub>Month=1</sub> | (%) Johnson<sub>Month=1</sub>  | (%) Pfizer<sub>Month=2</sub> | (%) Moderna<sub>Month=2</sub> | (%) Johnson<sub>Month=2</sub> | (%) Pfizer<sub>Month=3</sub> | (%) Moderna<sub>Month=3</sub> | (%) Johnson<sub>Month=3</sub> |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| Young       | 10 | 10 | 80 | 40 | 40 | 20 | 30 | 30 | 40 |
| Middle Aged | 10 | 80 | 10 | 20 | 60 | 20 | 40 | 20 | 40 |
| Old         | 80 | 10 | 10 | 50 | 20 | 30 | 10 | 20 | 70 |

Using this information, our model calculates the distribution of vaccines that need to be delivered to each destinations as (i.e.,  this matrix represents the vaccine demands of the destination sites):

| Destinations | Pfizer<sub>Month=1</sub> | Moderna<sub>Month=1</sub> | Johnson<sub>Month=1</sub>  | Pfizer<sub>Month=2</sub> | Moderna<sub>Month=2</sub> | Johnson<sub>Month=2</sub> | Pfizer<sub>Month=3</sub> | Moderna<sub>Month=3</sub> | Johnson<sub>Month=3</sub> |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| New York City | 360  | 727  | 413  | 465  | 705  | 330  | 473  | 338  |  690 |
| Los Angeles   | 1085 | 1820 | 595  | 1085 | 1610 | 805  | 1050 | 735  | 1715 |
| Chicago       | 2850 | 2850 | 1800 | 2700 | 3000 | 1800 | 1950 | 1650 | 3900 |

#### 2.1.b. Employees ####

There are four types of employees: junior, mid level, senior,  and principal, with the following wage distributions (values are in dollars and were synthetically generated):

| Employee Position | Hiring Cost | Firing Cost |
| --- | --- | --- |
| Junior     | \$2000 | \$300  |
| Mid Level  | \$3500 | \$500  |
| Senior     | \$4500 | \$800  |
| Principal  | \$7000 | \$1500 |

Each scientist can only work a certain number of hours in a month (values are in hours and were synthetically generated):

| Employee Position | Workable Hours | 
| --- | --- | 
| Junior     | 100 |
| Mid Level  | 250 |
| Senior     | 300 |
| Principal  | 450 |

Initailly, the distribution of employees at each of the production sites is (values are in number of employees and were synthetically generated): 

|  Site | Junior | Mid Level | Senior | Principal | 
| --- | --- | --- | --- | --- |
| Kalamazoo   | 15 | 7 | 3 | 1 |
| Bloomington | 10 | 8 | 8 | 0 |
| Baltimore   | 25 | 5 | 3 | 1 |

Each site has a limitation on the number of employees that they can hire for a certain position (values are in number of employees and were synthetically generated): 

|  Site | Junior | Mid Level | Senior | Principal | 
| --- | --- | --- | --- | --- |
| Kalamazoo   | 15000 | 8000 | 4000 | 500 |
| Bloomington | 18000 | 6000 | 4500 | 450 |
| Baltimore   | 19000 | 9500 | 2000 | 200 |

### 2.2. The Formal Model ###
Here we will layout the data, variables, constraints, and objective of our optimization model:

#### 2.2.1 Data
| Data | Explanation |
|  :--- | :--- |
| SRCS | set of ingredient sourcing locations $\in$ *{Andover, McPherson, Sanford}* |
| SITES | set of vaccine production locations $\in$ *{Kalamazoo, Bloomington, Baltimore}* |
| DESTS | set of vaccine destination locations $\in$ *{New York City, Los Angeles, Chicago}* |
| V | the type of vaccine $\in$ *{Pfizer, Moderna, Johnson}* |
| O | set of months that we are looking $\in$ *{1, 2, 3}* | 
| T | the methods of transportation $\in$ *{air, road, train}* |
| I | the ingredients needed to make a dose of vaccine $\in$ *{ing1, ing2, ing3}* |
| S | set of types of scientists $\in$ *{junior, mid_level, senior, principal}* |
|ing_supply<sub><b>nio</b></sub> | the amount of ingredient **i** located at source **n** in month **o** | 
|recipe<sub><b>vi</b></sub> | the amount of ingredient **i** needed to make vaccine **v** |
|demand<sub><b>nvo</b></sub> | the amount of vaccine **v** needed in destination **n** by the end of month **o** |
|ing_size<sub><b>it</b></sub> | the amount of ingredient **i** that can be shipped via method **t** in a single shipment |
|vac_size<sub><b>vt</b></sub> | the amount of vaccine **v** that can be shipped via method **t** in a single shipment |
|ing_ship_u<sub><b>ito</b></sub> | the amount of shipments of ingredient **i** that can be shipped via method **t** in month **o** |
|vac_ship_u<sub><b>vto</b></sub> | the amount of shipments of vaccine **v** that can be shipped via method **t** in month **o** |
|emp_u<sub><b>ns</b></sub> | the maximum number of employees of type **s** that can work at site **n** |
|vac_hours<sub><b>v</b></sub> | the number of worker hours needed to produce vaccine **v** |
|sci_hours<sub><b>s</b></sub> | the number of hours employee of type **s** can work in a month |
|ing_ship_c<sub><b>itn</b></sub> | the cost to ship ingredient **i** using method **t** from source **n** per shipment of ingredient |
|vac_ship_c<sub><b>vtn</b></sub> | the cost to ship vaccine **v** using method **t** from site **n** per shipment of vaccine |
|dist1<sub><b>nm</b></sub> | the distance from source **n** to site **m** |
|dist2<sub><b>nm</b></sub> | the distance from site **n** to dest **m** |
|vac_store_c<sub><b>nv</b></sub> | the cost to store vaccine **v** at site **n** |
|ing_store_c<sub><b>ni</b></sub> | the cost to store ingredient **i** at site **n** |

#### 2.2.2 Variables

| Variables | Explanation |
|  :--- | :--- |
|x<sub><b>ntmio</b></sub> | the amount of ingredient **i** shipped to from source **n** to site **m** via method **t** in month **o** |
|y<sub><b>nvo</b></sub> | the amount of vaccine **v** produced at site **n** in month **o** |
|z<sub><b>ntmvo</b></sub> | the amount of vaccine **v** shipped to from site **n** to destination **m** via method **t** in month **o** |
|ing<sub><b>nio</b></sub> | the amount of ingredient **i** stored at site **n** at the end of month **o** (for month 0, all values are 0 |
|vac<sub><b>nvo</b></sub> | the amount of vaccine **v** stored at site **n** at the end of month **o** (for month 0, all values are 0) |
|emp<sub><b>nso</b></sub> | the number of scientists of type **s** employed at site **n** in month **o** (for month 0, all values were predetermined) |
|h<sub><b>nso</b></sub> | the number of scientists of type **s** hired at site **n** in month **o** |
|f<sub><b>nso</b></sub> | the number of scientists of type **s** fired at site **n** in month **o** |
|h_c<sub><b>nso</b></sub> | the cost to hire a scientist of type **s** |
|f_c<sub><b>nso</b></sub> | the cost to fire a scientist of type **s** |

#### 2.2.3 Constraints

| Description | Constraint |
|  :--- | :--- |
|Cannot ship more ingredients than we have | $$\sum_{t \in T} \sum_{m \in SITES} x_{ntmio} \leq ing\_supply_{nio} \space \forall n \in SRCS, i \in I, o \in O $$ |
|Inventory balance for ingredients | $$ \sum_{t \in T} \sum_{n \in SRCS} (x_{ntmio} + ing_{ntmi(o-1)}) = \sum_{v \in V} (y_{nvo} \cdot recipe_{vi}) + ing_{ntmio} \space \forall m \in SITES, i \in I, o \in O $$ |
|Must at least meet demands for vaccines | $$ \sum_{n \in SITES}  (y_{nvo} + vac_{nvo}) \geq demand_{mvo} \space \forall m \in DESTS, v \in V, o \in O $$ |
|Inventory balance for vaccines | $$ y_{nvo} + vac_{nv(o-1)} = \sum_{t \in T} \sum_{m \in DESTS} z_{ntmvo} + vac_{nvo}  \space \forall n \in SITES, v \in V, o \in O $$|
|Ship exactly the amount of vaccines that are needed to meet demands | $$\sum_{n \in SITES} \sum_{t \in T} \sum_{m \in DESTS} z_{ntmvo} = \sum_{m \in DESTS} demand_{mvo}  \space \forall v \in V, o \in O $$|
|Limit ingredient shipment by transportation method capacity | $$ \sum_{n \in SRCS} \sum_{m \in SITES} (x_{ntmio} / ing\_size_{it}) \leq  ing\_ship\_u_{ito}  \space \forall i \in I, t \in T, o \in O $$|
|Limit vaccine shipment by transportation method capacity | $$ \sum_{n \in SITES} \sum_{m \in DESTS} (z_{ntmvo} / vac\_size_{vt}) \leq  vac\_ship\_u_{vto}  \space \forall v \in V, t \in T, o \in O $$|
|Employee balance | $$ emp_{ns(o-1)} + h_{nso} + f_{nso} = emp_{nso} \space \forall n \in SITES, s \in S, o \in O $$ |
|Work hours balance | $$ y_{nvo} * vac\_hours_{v} \leq \sum_{s \in S} emp_{nso} \cdot sci\_hours_{s} \space \forall n \in SITES, v \in V, o \in O $$ |
|Limit number of employees | $$ emp_{nso} \leq emp\_u_{ns} \space \forall n \in SITES, s \in S, o \in O $$|

#### 2.2.4 Objective

The objective is to minimize the sum of the cost to ship ingredients, ship vaccines, store ingredients, store vaccines, hire employees, and fire employees

$$
\min_{x, y, z, ing, vac, emp, h, f} \sum_{i \in I} \sum_{t \in T} \sum_{o \in O} \sum_{s \in S} \sum_{l \in SRCS} \sum_{n \in SITES} \sum_{m in DESTS} (ing\_ship\_c_{itl} \cdot dist1_{ln} \cdot x_{ntmio} \div ing\_size_{it}) + (vac\_ship\_c_{vtn} \cdot dist2_{nm} \cdot z_{ntmvo} \div vac\_size_{vt}) + (ing\_store\_c_{ni} \cdot ing_{nio}) + (vac\_store\_c_{nv} \cdot vac_{nvo}) + (h_{nso} \cdot h\_c_{s}) + (f_{nso} \cdot f\_c_{s})
$$

### 2.3. Standard Form ###

Do to the sheer size of our model (675 variables and 252 constraints), it was infeasible to write out this model in standard form. 

<!-- However, since the program is linear, the following steps "could" be applied to transform the formal model into standard form:

#### Step 1. Objective ####
Convert the min problem to a max problem by utilizing the property that :
$ 
\begin{align}
\min_{x} c^{T} \cdot x = -\max_{x} (-c^{T}x)
\end{align}
$

#### Step 2. Contraint 1 ####
No change needs to be made to constraint 1, since it is already in the form: 
$ 
\begin{align}
Ax \leq b
\end{align}
$

#### Step 3. Constraint 2 #### -->


### 2.4. Modelling Details ###
We chose to model the problem of maufacturing, storing, and distributing the three types of the COVID-19 vaccine within a single time period as a Minimum-Cost Network Flow (MCNF) optimizaiton problem. The problem was then expanded across a three month time period, thus, including features of a Multi-Period (MP) Transportation Problem (TP). 

A MCNF models a problem as a directed graph, allowing the amount of material (in this case, ingredients and vaccines) to move along certain arcs (directed edges between nodes in the graph). This model optimizes for providing the best routing scheme to get all required material from source nodes to site nodes (production sites) to destination nodes. Since the the COVID-19 vaccine manufacturing and distribution process is an archetype of such a use case, the MNCF model matches the nuances of this problem well. 

When we incorporated time, the problem took on additional features that are common to MP-TPs, which allowed the model to incorporate the possibility of storing ingredients and vaccines at different locations. This matches the real world use case, since the production and delivery of COVID-19 vaccines is a problem that spans beyond a single instance (i.e., multiple months). 

## 3. Solution ##

We first begin by creating arrays to store the data that we need.

In [177]:
########################################################################
##                                                                    ##
##                             Data                                   ##
##                                                                    ##
########################################################################

# these are the available shipment methods
shipment_methods = [:air, :road, :train]

# Flow through the network should go from sources -> sites -> destinations

# these are the locations from which we can procure the ingredients. 
sources = [:Andover, :McPherson, :Sanford]

# these are the locations where vaccines are produced
sites = [:Kalamazoo, :Bloomington, :Baltimore]  

# these are the locations where vaccines need to be delivered
destinations = [:NewYorkCity, :LosAngeles, :Chicago] 

# these are the different types of vaccines available
vaccine_types = [:pfizer, :moderna, :johnson]

# these are the different population categories in each city
category = [:young, :middle_aged, :old]

# these are the different ingredient types available
ingredients = [:ing1, :ing2, :ing3]

# these are the different types of scientist
scientists = [:junior, :mid_level, :senior, :principal]

# these are the different months for which we need to plan
months = [1,2,3]

using NamedArrays

########################################################################
##                                                                    ##
##                             Synthetic Data                         ##
##                                                                    ##
########################################################################

# distances from ingredient sources to production sites
dist1 = NamedArray([50 25 84; 12 75 18;64 39 70], (sources, sites), ("Sources","Sites") )
println(dist1)
println("\n======================\n")

# distances from production sites to destinations
dist2 = NamedArray([40 15 74; 22 85 28;44 19 50], (sites, destinations), ("Sites","Destinations"))
println(dist2)
println("\n======================\n")

# age distribution of population in each city
distribution = NamedArray([0.25 0.55 0.2; 0.1 0.6 0.3;0.2 0.4 0.4], 
    (destinations, category), ("Destinations","Category"))
println(distribution)
println("\n======================\n")

# vaccine requirement for each population type by month
vaccine_req = zeros(3, 3, 3)
vaccine_req[:, :, 1] = [0.1 0.1 0.8; 0.1 0.8 0.1; 0.8 0.1 0.1]
vaccine_req[:, :, 2] = [0.4 0.4 0.2; 0.2 0.6 0.2; 0.5 0.2 0.3]
vaccine_req[:, :, 3] = [0.3 0.3 0.4; 0.4 0.2 0.4; 0.1 0.2 0.7]

vaccine_req_proportion_NA = NamedArray(vaccine_req, 
                        (category, vaccine_types, months), ("Category","Vaccine Types", "Months"))
println(vaccine_req_proportion_NA)
println("\n======================\n")

# Amount of ingredients available in each source
supply = zeros(3, 3, 3)
supply[:, :, 1] = [850000 622000 905000; 700000 615000 900000; 950000 820000 784000]
supply[:, :, 2] = [450400 142000 405000; 200050 241000 80000; 400600 608000 700000]
supply[:, :, 3] = [140000 182000 205000; 20000 321000 80000; 409000 201000 100450]

supply_NA = NamedArray(supply, (sources, ingredients, months), ("Sources", "Ingredients", "Months"))
println(supply_NA)
println("\n======================\n")

# Amount of ingredient needed for making each vaccine
demand_NA = NamedArray([16 20 9; 4 25 12; 7 14 35], 
    (vaccine_types, ingredients), ("Vaccine Type", "Ingredients"))

println(demand_NA)
println("\n======================\n")

# total population in each destination city
population = Dict(zip( destinations, [1500 3500 7500]))
println(population)
println("\n======================\n")

# cost of shipping ingredients
cost_of_shipping_ingredients = zeros(3, 3, 3);
cost_of_shipping_ingredients[:, :, 1] = [2 5 7; 4 8 12; 3 7 8];
cost_of_shipping_ingredients[:, :, 2] = [1 5 6; 2 4 9; 3 5 7];
cost_of_shipping_ingredients[:, :, 3] = [1 4 8; 3 6 7; 2 5 7];
cost_of_shipping_ingredients_NA =  NamedArray(cost_of_shipping_ingredients, 
    (sources, shipment_methods, ingredients), ("Sources", "Shipment Methods", "Ingredients"))
println(cost_of_shipping_ingredients_NA)
println("\n======================\n")

# Wages for different levels of scientists
cost_of_hiring_scientists = Dict(zip(scientists, [2000 3500 4500 7000]))

println(cost_of_hiring_scientists)
println("\n======================\n")

# Cost of firing different levels of scientists
cost_of_firing_scientists = Dict(zip(scientists, [300 500 800 1500]))
println(cost_of_firing_scientists)
println("\n======================\n")

# cost of shipping vaccines.
cost_of_shipping_vaccines = zeros(3, 3, 3);
cost_of_shipping_vaccines[:, :, 1] = [2 5 7; 4 8 12; 3 7 8];
cost_of_shipping_vaccines[:, :, 2] = [1 5 6; 2 4 9; 3 5 7];
cost_of_shipping_vaccines[:, :, 3] = [1 4 8; 3 6 7; 2 5 7];
cost_of_shipping_vaccines_NA =  NamedArray(cost_of_shipping_vaccines, 
    (sites, shipment_methods, vaccine_types), ("Sites", "Shipment Methods", "Vaccine Types"))
println(cost_of_shipping_vaccines_NA)
println("\n======================\n")

# ingredient shipment size. Amount of ingredient that can be sent in one shipment for each shipment mode.
ingredient_shipment_size_NA =  NamedArray([1200 5000 8200; 4200 6240 3540; 4710 5620 3840], 
    (ingredients, shipment_methods), ("Ingredients", "Shipment Methods"))
println(ingredient_shipment_size_NA)
println("\n======================\n")

# vaccine shipment size. Amount of vaccine that can be sent in one shipment for each shipment mode.
vaccine_shipment_size_NA =  NamedArray([1500 5000 840; 2000 6000 3580; 4200 5600 3400], 
    (vaccine_types, shipment_methods), ("Vaccines", "Shipment Methods"))
println(vaccine_shipment_size_NA)
println("\n======================\n")

# total number of shipments of each type that can be used for shipping ingredients in each month
number_of_ingredient_shipment_by_ship_method = zeros(3, 3, 3)
number_of_ingredient_shipment_by_ship_method[:, :, 1] = [30 90 200; 20 100 350; 25 30 80]
number_of_ingredient_shipment_by_ship_method[:, :, 2] = [15 45 800; 60 50 450; 80 100 600]
number_of_ingredient_shipment_by_ship_method[:, :, 3] = [65 95 100; 40 50 150; 90 300 800]
number_of_ingredient_shipments_NA = NamedArray(number_of_ingredient_shipment_by_ship_method, 
                (ingredients, shipment_methods, months), ("Ingredients", "Shipment methods", "Months"))
println(number_of_ingredient_shipments_NA)
println("\n======================\n")

# total number of shipments of each type that can be used for shipping vaccines in each month
number_of_vaccine_shipment_by_ship_method = zeros(3, 3, 3)
number_of_vaccine_shipment_by_ship_method[:, :, 1] = [120 240 370; 250 100 350; 215 130 840]
number_of_vaccine_shipment_by_ship_method[:, :, 2] = [101 280 340; 450 130 300; 245 430 440]
number_of_vaccine_shipment_by_ship_method[:, :, 3] = [128 244 373; 240 120 340; 255 330 240]
number_of_vaccine_shipments_NA = NamedArray(number_of_vaccine_shipment_by_ship_method, 
                (vaccine_types, shipment_methods, months), ("Vaccine Types", "Shipment methods", "Months"))
println(number_of_vaccine_shipments_NA)
println("\n======================\n")

# Number of working hours needed for making each vaccine
working_hours_for_vaccine = Dict(zip( vaccine_types, [140 210 340]))

println(working_hours_for_vaccine)
println("\n======================\n")

# Number of working hours from each scientist
scientist_working_hours = Dict(zip(scientists, [100 250 300 450]))
println(scientist_working_hours)
println("\n======================\n")

# Initial number of scientists in each site
initial_number_of_scientists_NA =  NamedArray([15 7 3 1; 10 8 8 0; 25 5 3 1], 
    (sites, scientists), ("Sites", "Scientists"))
println(initial_number_of_scientists_NA)
println("\n======================\n")

# Cost of storing ingredients in inventory in each site
ingredient_inventory_cost = NamedArray([120 200 210; 20 35 15; 124 300 150], 
                (sites, ingredients), ("Sites", "Ingredients"))
println(ingredient_inventory_cost)
println("\n======================\n")

# Cost of storing vaccines in inventory in each site
vaccine_inventory_cost = NamedArray([400 280 610; 250 325 165; 380 459 357], 
                (sites, vaccine_types), ("Sites", "Vaccine Types"))
println(vaccine_inventory_cost)
println("\n======================\n")

# Maximum number of employees of each type that can be hired in each site in every month
max_employees_per_site_NA = NamedArray([15000 8000 4000 500; 18000 6000 4500 450; 19000 9500 2000 200], 
                (sites, scientists), ("Sites", "Scientists"))
println(max_employees_per_site_NA)
println("\n======================\n")

3×3 Named Matrix{Int64}
Sources ╲ Sites │   :Kalamazoo  :Bloomington    :Baltimore
────────────────┼─────────────────────────────────────────
:Andover        │           50            25            84
:McPherson      │           12            75            18
:Sanford        │           64            39            70


3×3 Named Matrix{Int64}
Sites ╲ Destinations │ :NewYorkCity   :LosAngeles      :Chicago
─────────────────────┼─────────────────────────────────────────
:Kalamazoo           │           40            15            74
:Bloomington         │           22            85            28
:Baltimore           │           44            19            50


3×3 Named Matrix{Float64}
Destinations ╲ Category │       :young  :middle_aged          :old
────────────────────────┼─────────────────────────────────────────
:NewYorkCity            │         0.25          0.55           0.2
:LosAngeles             │          0.1           0.6           0.3
:Chicago                │          0.2 

In [178]:
########################################################################
##                                                                    ##
##                             Derived Data                           ##
##                                                                    ##
########################################################################

# Get population count for each destination city
population_sizes = zeros(size(destinations)[1], size(category)[1])
population_sizes_NA = NamedArray(population_sizes, 
                        (destinations, category), ("Destinations","Category"))

for x in destinations
    for y in category
        population_sizes_NA[x, y] = distribution[x ,y] * population[x]
    end
end

println(population_sizes_NA)
println("\n======================\n")

# Get vaccine count for each destination city
vaccine_reqs = ones(size(destinations)[1], size(vaccine_types)[1], size(months)[1])
vaccine_reqs_NA = NamedArray(vaccine_reqs, 
                        (destinations, vaccine_types, months), ("Destinations","Vaccine Types", "Months"))


for x in destinations
    for y in vaccine_types
        for z in months
            vaccine_reqs_NA[x, y, z] = sum(population_sizes_NA[x, w] * 
                                    vaccine_req_proportion_NA[w, y, z] for w in category)
        end
    end
end


println(vaccine_reqs_NA)
println("\n======================\n")

3×3 Named Matrix{Float64}
Destinations ╲ Category │       :young  :middle_aged          :old
────────────────────────┼─────────────────────────────────────────
:NewYorkCity            │        375.0         825.0         300.0
:LosAngeles             │        350.0        2100.0        1050.0
:Chicago                │       1500.0        3000.0        3000.0


3×3×3 Named Array{Float64, 3}

[:, :, Months=1] =
Destinations ╲ Vaccine Types │  :pfizer  :moderna  :johnson
─────────────────────────────┼─────────────────────────────
:NewYorkCity                 │    360.0     727.5     412.5
:LosAngeles                  │   1085.0    1820.0     595.0
:Chicago                     │   2850.0    2850.0    1800.0

[:, :, Months=2] =
Destinations ╲ Vaccine Types │  :pfizer  :moderna  :johnson
─────────────────────────────┼─────────────────────────────
:NewYorkCity                 │    465.0     705.0     330.0
:LosAngeles                  │   1085.0    1610.0     805.0
:Chicago                   

In [179]:
using JuMP, Clp

m = Model(Clp.Optimizer)

########################################################################
##                                                                    ##
##                             Variables                              ##
##                                                                    ##
########################################################################

# x[i,j,k,l,m] is the amount of ingredient "l" sent from source i to site k using shipment method j in month m
@variable(m, x[sources, shipment_methods, sites, ingredients, months] >= 0)

# y[i,j,k] is the number of vaccines produced in site i of type j in month k
@variable(m, y[sites, vaccine_types, months] >= 0)

# z[i,j,k,l,m] is the number of vaccines used to meet demand from site i 
# to destination k of type "l" using shipment method j in month m
@variable(m, z[sites, shipment_methods, destinations, vaccine_types, months] >= 0)

# emp[i, j, k] is the number of scientists of type j employed in site i in month k
@variable(m, emp[sites, scientists, months] >= 0)

# h[i, j, k] is the number of scientists of type j hired in site i in month k
@variable(m, h[sites, scientists, months] >= 0)

# f[i, j, k] is the number of scientists of type j fired in site i in month k
@variable(m, f[sites, scientists, months] >= 0)

# amount of each ingredient left at the end of each month (in inventory) in each site
@variable(m, ing[sites, ingredients, months] >= 0)

# amount of each vaccine type left at the end of each month (in inventory) in each site
@variable(m, vac[sites, vaccine_types, months] >= 0)

########################################################################
##                                                                    ##
##                             Constraints                            ##
##                                                                    ##
########################################################################

# number of ingredients supplied from source i should be less than the total available across all months
@constraint(m, sup[i in sources, l in ingredients, n in months], 
                                sum(x[i,j,k,l,n] for j in shipment_methods, k in sites) 
                                        <= supply_NA[i, l, n]);

# INVENTORY BALANCE FOR INGREDIENTS IN EACH SITE #
# amount we start with + amount we procure = amount we use + amount that carries to next month

# Initial inventory balance for ingredients
@constraint(m, ing_inv_balance_init[k in sites, l in ingredients, q in months[1:1]],
         sum(x[i, j, k, l, q] for i in sources, j in shipment_methods) == 
            sum(y[k,v,q] * demand_NA[v, l] for v in vaccine_types) + ing[k, l, q]);

# Inventory balance for the remaining months
@constraint(m, ing_inv_balance[k in sites, l in ingredients, q in months[2:size(months)[1]]],
                sum(x[i, j, k, l, q] for i in sources, j in shipment_methods) + ing[k, l, q-1] == 
                    sum(y[k,v,q] * demand_NA[v, l] for v in vaccine_types) + ing[k, l, q]);

# Sum of vaccines produced in all sites should be greater than or equal to demand in the first month.
@constraint(m, demand_cons_init[l in vaccine_types, k in destinations, q in months[1:1]],
                                sum(y[i, l, q] for i in sites) >= vaccine_reqs_NA[k, l, q]);

# Sum of vaccines produced in all sites and inventory from previous month
# should be greater than or equal to demand in the subsequent months.
@constraint(m, demand_cons[l in vaccine_types, k in destinations, q in months[2:size(months)[1]]],
                            sum(y[i, l, q] for i in sites) + sum(vac[i, l, q-1] for i in sites) >= 
                                    vaccine_reqs_NA[k, l, q]);

# INVENTORY BALANCE FOR VACCINES IN EACH SITE
# amount we start with + amount we produce  = amount we deliver + amount that carries to next month
# Initial inventory balance for vaccines
@constraint(m, vac_inv_balance_init[i in sites, l in vaccine_types, q in months[1:1]],
                                y[i, l, q] == sum(z[i, j, k, l, q] for j in shipment_methods, k in destinations) + 
                                        vac[i, l, q]);

# Inventory balance for the remaining months
@constraint(m, vac_inv_balance[i in sites, l in vaccine_types, q in months[2:size(months)[1]]],
                        y[i, l, q] + vac[i, l, q-1] == 
                                    sum(z[i, j, k, l, q] for j in shipment_methods, k in destinations) + 
                                        vac[i, l, q]);

# Total amount of vaccine used to meet demand should be equal to the total demand across destinations
@constraint(m, demand_constraint[l in vaccine_types, q in months], 
                sum(z[i, j, k, l, q] for i in sites, j in shipment_methods, k in destinations) 
                        == sum(vaccine_reqs_NA[k, l, q] for k in destinations))


# For ingredients - number of shipments by each shipment method should be bound by the total allowed. 
@constraint(m, ing_shipment[j in shipment_methods, l in ingredients, q in months],
                    sum(x[i, j, k, l, q] for i in sources, k in sites) / ingredient_shipment_size_NA[l, j] 
                        <= number_of_ingredient_shipments_NA[l, j, q]);

# For vaccines - number of shipments by each shipment method should be bound by the total allowed. 
@constraint(m, vac_shipment[j in shipment_methods, l in vaccine_types, q in months],
                    sum(z[i, j, k, l, q] for i in sites, k in destinations) / vaccine_shipment_size_NA[l, j] 
                        <= number_of_vaccine_shipments_NA[l, j, q]);

# SCIENTISTS BALANCE #
# number currently employed - number fired + number hired = number workers available

# balance scientists in first month
@constraint(m, sci_bal_init[i in sites, j in scientists, q in months[1:1]], 
            initial_number_of_scientists_NA[i, j] + h[i, j, q] - f[i, j, q] == emp[i, j, q]);

# balance scientists in all months after the first
@constraint(m, sci_bal[i in sites, j in scientists, q in months[2:size(months)[1]]], 
            emp[i, j, q-1] + h[i, j, q] - f[i, j, q] == emp[i, j, q]);

# number of vaccines produced should be less than or equal to the total working hours available from
# all employees
@constraint(m, hour_constraint[i in sites, l in vaccine_types, q in months],
            y[i, l, q] * working_hours_for_vaccine[l] 
                        <= sum(emp[i, j, q] * scientist_working_hours[j] for j in scientists)); 

# number of employees should be bound by the max allowed
@constraint(m, max_employee[i in sites, j in scientists, q in months],
                emp[i, j, q] <= max_employees_per_site_NA[i, j]);

########################################################################
##                                                                    ##
##                             Expressions                            ##
##                                                                    ##
########################################################################

# Cost of shipping ingredients from sources to sites
@expression(m, ingredient_shipping_cost[l in ingredients, j in shipment_methods], 
              cost_of_shipping_ingredients_NA[i, j, l] * dist1[i, k] *
                (sum(x[i,j,k,l,q] for i in sources, k in sites, q in months) / ingredient_shipment_size_NA[l, j]));

# Cost of shipping vaccines from sites to destinations
@expression(m, vaccine_shipping_cost[l in vaccine_types, j in shipment_methods], 
        cost_of_shipping_vaccines_NA[i, j, l] * dist2[i, k] *
         (sum(z[i,j,k,l,q] for i in sites, k in destinations, q in months) / vaccine_shipment_size_NA[l, j]));


# Ingredient storage cost
@expression(m, ingredient_storage_cost[k in months], sum(ingredient_inventory_cost[i, j] * ing[i, j, k] 
                               for i in sites, j in ingredients));

# Vaccine storage cost
@expression(m, vaccine_storage_cost[k in months], sum(vaccine_inventory_cost[i, j] * vac[i, j, k] 
                        for i in sites, j in vaccine_types));

# Cost of hiring scientists
@expression(m, hiring_cost[k in months], sum(sum(h[i, j, k] for i in sites) * cost_of_hiring_scientists[j]
                                            for j in scientists));

# Cost of firing scientists
@expression(m, firing_cost[k in months], sum(sum(f[i, j, k] for i in sites) * cost_of_firing_scientists[j]
                                            for j in scientists));

########################################################################
##                                                                    ##
##                             Objective                              ##
##                                                                    ##
########################################################################

# Minimize ingredient_shipping_cost and vaccine_shipping_cost
@objective(m, Min, sum(ingredient_shipping_cost[l, j] for l in ingredients, j in shipment_methods) + 
                   sum(vaccine_shipping_cost[l, j] for l in vaccine_types, j in shipment_methods) +
                   sum(ingredient_storage_cost[k] for k in months) + 
                   sum(vaccine_storage_cost[k] for k in months) +
                   sum(hiring_cost[k] for k in months) + 
                   sum(firing_cost[k] for k in months));

########################################################################
##                                                                    ##
##                             Optimize                               ##
##                                                                    ##
########################################################################
optimize!(m);

Coin0506I Presolve 212 (-58) rows, 502 (-173) columns and 1434 (-660) elements
Clp0006I 0  Obj 0 Primal inf 74682.001 (29)
Clp0006I 56  Obj 7952.1747 Primal inf 164271.39 (49)
Clp0006I 118  Obj 15304.023 Primal inf 197186.06 (31)
Clp0006I 161  Obj 19497969
Clp0000I Optimal - objective value 19497969
Coin0511I After Postsolve, objective 19497969, infeasibilities - dual 0 (0), primal 0 (0)
Clp0032I Optimal objective 19497968.82 - 161 iterations time 0.002, Presolve 0.00


In [180]:
# Print igredient information
for o in months
    for i in ingredients
        for n in sources
            for m in sites
                for t in shipment_methods
                    if value(x[n,t,m,i,o]) > 0.0
                        println("Shipped ",value(x[n,t,m,i,o]), " of ",i, " via ",t, " from ",n," to ",m," in month ", o)
                    end
                end
            end
        end
    end
end

for n in sites
    for o in months
        for i in ingredients
            if value(ing[n,i,o]) > 0.0
                println("Stored ",value(ing[n,i,o]), " of ",i," at ",n," in month ", o)
            end
        end
    end
end

# Print vaccine information 
for o in months
    for v in vaccine_types
        for n in sites
            for m in destinations
                for t in shipment_methods
                    if value(z[n,t,m,v,o]) > 0.0
                        println("Shipped ",value(z[n,t,m,v,o]), " of ",v, " via ",t, " from ",n," to ",m," in month ", o)
                    end
                end
            end
        end
    end
end
for o in months
    for n in sites
        for v in vaccine_types
            if value(y[n, v, o]) > 0.0
                println("Produced ",value(y[n,v,o]), " of ",v, " at ",n, " in month ", o)
            end
        end
    end
end
for n in sites
    for o in months
        for v in vaccine_types
            if value(vac[n,v,o]) > 0.0
                println("Stored ",value(vac[n,v,o]), " of ",v," at ",n," in month ", o)
            end
        end
    end
end

# Print employee information
for o in months
    for n in sites
        for s in scientists
            println("Currently employed ", value(emp[n,s,o]), " ", s, "'s at ", n, " in month ", o)
            println("Hired  ", value(h[n,s,o]), " ", s, "'s at ", n, " in month ", o)
            println("Fired  ", value(f[n,s,o]), " ", s, "'s at ", n, " in month ", o)
        end
    end
end

Shipped 23139.00276806527 of ing1 via train from Andover to Bloomington in month 1
Shipped 95176.83056526804 of ing1 via train from McPherson to Kalamazoo in month 1
Shipped 104.99999999999997 of ing1 via train from McPherson to Baltimore in month 1
Shipped 83999.99999999999 of ing2 via air from Andover to Bloomington in month 1
Shipped 192849.16666666663 of ing2 via road from McPherson to Kalamazoo in month 1
Shipped 209.99999999999997 of ing2 via road from McPherson to Baltimore in month 1
Shipped 97943.48703379954 of ing3 via air from Andover to Bloomington in month 1
Shipped 19281.51296620046 of ing3 via air from McPherson to Kalamazoo in month 1
Shipped 126229.16666666666 of ing3 via road from McPherson to Kalamazoo in month 1
Shipped 524.9999999999999 of ing3 via air from McPherson to Baltimore in month 1
Shipped 27459.16149822401 of ing1 via train from Andover to Bloomington in month 2
Shipped 89806.67183510931 of ing1 via train from McPherson to Kalamazoo in month 2
Shipped 104

## 4. Results and Discussion ##

### 4.1. Ingredients ###

From the output above, we can see that in month 1 we make the following shipments:

| Ingredient | Arc | mL of Ingredient Shipped |
| --- | --- | --- |
| Ing1 | Andover--<sub>train</sub>->Bloomington |   23,139 |
| Ing1 | McPherson--<sub>train</sub>->Kalamazoo |   95,177 |
| Ing1 | McPherson--<sub>train</sub>->Baltimore |   105 |
| Ing2| Andover--<sub>air</sub>->Bloomington   |   84,000 |
| Ing2|McPherson--<sub>road</sub>->Kalamazoo  |   192,850 |
| Ing2|McPherson--<sub>road</sub>->Baltimore  |   210 |
| Ing3|Andover--<sub>air</sub>->Bloomington   |   97,944 |
| Ing3|McPherson--<sub>air</sub>->Kalamazoo   |   19,282 |
| Ing3|McPherson--<sub>road</sub>->Kalamazoo  |   126,230 |
| Ing3|McPherson--<sub>air</sub>->Baltimore   |   525 |

Since a part of the cost is calculated based off of a distance model, we would expect the arcs from sources to sites to be consistent, and that is exactly what we see in month 1. However, we see that the shipment method is **not** consistent for the various ingredients. This tells us that the model is very sensitive to the changes in the shipment method costs. 

In Month 2 we make the following shipments:

| Ingredient | Arc | mL of Ingredient Shipped |
| --- | --- | --- |
| Ing1 | Andover--<sub>train</sub>->Bloomington |   27,460 |
| Ing1 | McPherson--<sub>train</sub>->Kalamazoo |   89,807 |
| Ing1 | McPherson--<sub>train</sub>->Baltimore |   105 |
| Ing2| Andover--<sub>air</sub>->Bloomington   |   111,001 |
| Ing2|McPherson--<sub>air</sub>->Kalamazoo  |   140,790 |
| Ing2|McPherson--<sub>road</sub>->Kalamazoo  |   22,097 |
| Ing2|McPherson--<sub>air</sub>->Baltimore  |   210 |
| Ing3|Andover--<sub>air</sub>->Bloomington   |   110,904 |
| Ing3|Andover--<sub>air</sub>->Kalamazoo   |   51,681 |
| Ing3|McPherson--<sub>air</sub>->Kalamazoo  |   79,475 |
| Ing3|McPherson--<sub>air</sub>->Baltimore   |   525 |


In month 3 we make the following shipments:

| Ingredient | Arc | mL of Ingredient Shipped |
| --- | --- | --- |
| Ing1 | Andover--<sub>train</sub>->Bloomington |   74,561 |
| Ing1 | McPherson--<sub>train</sub>->Kalamazoo |   19,895 |
| Ing1 | McPherson--<sub>train</sub>->Baltimore |   105 |
| Ing2| Andover--<sub>air</sub>->Bloomington   |   158,349 |
| Ing2|McPherson--<sub>air</sub>->Kalamazoo  |   9,442 |
| Ing2|McPherson--<sub>road</sub>->Kalamazoo  |   25,735 |
| Ing2|McPherson--<sub>road</sub>->Baltimore  |   210 |
| Ing3|Andover--<sub>air</sub>->Bloomington   |   131,778 |
| Ing3|McPherson--<sub>air</sub>->Kalamazoo   |   72,174 |
| Ing3|McPherson--<sub>air</sub>->Baltimore   |   525 |

Another interesting thing to note is that we never store any of the ingredients at any of the sites in any of the months. This is because it is must cheaper to simply ship the amount of ingredient on an as needed basis. This tells us that our constraints  for shipping ingredients are non-binding, i.e., that we can ship more ingredients if need be, and are not limited by the shipping amount limitations.

### 4.2. Vaccines ###

Looking at the output above, we can see how much of each vaccine were produced at the production sites:

| Sites | mL of Pfizer<sub>Month=1</sub> | mL of Moderna<sub>Month=1</sub> | mL of Johnson<sub>Month=1</sub> |  mL of Pfizer<sub>Month=2</sub> | mL of Moderna<sub>Month=2</sub> | mL of Johnson<sub>Month=2</sub> | mL of Pfizer<sub>Month=3</sub> | mL of Moderna<sub>Month=3</sub> | mL of Johnson<sub>Month=3</sub> |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| Kalamazoo | 4295 | 3179 | 1964 | 4250 | 2017 | 1964 | 385 | 0 | 1964 |
| Bloomington | 0 | 2219 | 2038 | 0 | 3299 | 2038 | 3089 | 2723 | 2038 |
| Baltimore | 0 | 0 | 15 | 0 | 0 | 15 | 0 | 0 | 15 |

Additionally, we see that only 1209 mL and 2290 mL of the Johnson vaccines are stored at the Bloomington site in months 1 and 2, respectively.

Finally, the data below shows which arcs were used in order to ship the vaccines to destination nodes:

In month 1:

| Vaccine | Arc | mL of Vaccine Shipped |
| --- | --- | --- |
| Pfizer | Kalamazoo--<sub>road</sub>->Los Angeles |   4295 |
| Moderna | Kalamazoo--<sub>air</sub>->Los Angeles |   3179 |
| Moderna | Bloomington--<sub>road</sub>->New York City |   2219 |
| Johnson| Kalamazoo--<sub>air</sub>->Los Angeles   |   1964 |
| Johnson|Bloomington--<sub>air</sub>-> New York City |   830 |
| Johnson|Baltimore--<sub>air</sub>->Los Angeles  |   15 |

In month 2:

| Vaccine | Arc | mL of Vaccine Shipped |
| --- | --- | --- |
| Pfizer | Kalamazoo--<sub>road</sub>->Los Angeles |   4250 |
| Moderna | Kalamazoo--<sub>air</sub>->Los Angeles |   2017 |
| Moderna | Bloomington--<sub>road</sub>->New York City |   3299 |
| Johnson| Kalamazoo--<sub>air</sub>->Los Angeles   |   1964 |
| Johnson|Bloomington--<sub>air</sub>-> New York City |   957 |
| Johnson|Baltimore--<sub>air</sub>->Los Angeles  |   15 |

In month 3:

| Vaccine | Arc | mL of Vaccine Shipped |
| --- | --- | --- |
| Pfizer | Kalamazoo--<sub>road</sub>->Los Angeles |   386 |
| Pfizer | Bloomington--<sub>road</sub>->New York City |   3088 |
| Moderna | Bloomington--<sub>road</sub>->New York City |   2723 |
| Johnson| Kalamazoo--<sub>air</sub>->Los Angeles   |   1964 |
| Johnson|Bloomington--<sub>air</sub>-> New York City |   4327 |
| Johnson|Baltimore--<sub>air</sub>->Los Angeles  |   15 |

It's interesting to note that, regardless of the month, usually the same arcs (site to destination pair) is chosen. This indicates that the model is heavily influenced by the distance matrix that outlines the distances between the set of sites and the set of destinations. 

### 4.3. Employess ###

From the above output, we can see the flow of workers at the production sites:

At Kalamazoo:

| Scientist Type | # Employed Initially | # Employed Month 1 | # Employed Month 2 | # Employed Month 3 |
| --- | --- | --- | --- | --- |
| Junior    | 15 | 15   | 15   | 15   |
| Mid Level | 7  | 2659 | 2659 | 2659 |
| Senior    | 3  | 3    | 3    | 3    |
| Principal | 1  | 1    | 1    | 1    |

At Bloomington:

| Scientist Type | # Employed Initially | # Employed Month 1 | # Employed Month 2 | # Employed Month 3 |
| --- | --- | --- | --- | --- |
| Junior    | 10 | 10   | 10   | 10   |
| Mid Level | 8  | 2758 | 2758 | 2758 |
| Senior    | 8  | 8    | 8    | 8    |
| Principal | 0  | 0    | 0    | 0    |

At Baltimore:

| Scientist Type | # Employed Initially | # Employed Month 1 | # Employed Month 2 | # Employed Month 3 |
| --- | --- | --- | --- | --- |
| Junior    | 25 | 25 | 25 | 25   |
| Mid Level | 5  | 5  | 5  | 5 |
| Senior    | 3  | 3  | 3  | 3    |
| Principal | 1  | 1  | 1  | 1 |

What's interesting to see here is that at the Kalamazoo and and Bloomington locations, we don't start out with enough employees to meet vaccine manufacturing demands in a cost effective way. So at these two locations, there is an influx of employees. Further, we see that in the first month (when the influx occurs), only mid-level type scientists are hired. This tells us that the mid-level type scientist are the most cost effective workers in the job market. Additionally, we see no change in the distribution of workers at the Baltimore location. This can be attributed to this location starting out with a large number of employees (34), yet only needing to produce 15 mL of the Johnson vaccine every month (see 4.3 for further clarification). This tells us that this plant is most likely overstaffed.


### 4.4. Limitations ###

One limitation of the model is that we didn't quite have the time to incorporate the salary for the scientists. In this case, it is always advantageous to hire more employees when the demands require it, since there is no associated cost to maintain an employee once hired. This results in the model never proposing to fire anyone.

## 5. Conclusion ##

### 5.1. Summary ###

In summary, our model estimates that it will cost ~\$19,497,969 to ship ingredients, manufacture vaccines, and deliver vaccines over the three month period to meet vaccination demands. While this number bears little meaning since the values chosen were the data section were artificial, it shows the succintness of the model to be able to collapse a variety of information into easy to interpret results. 

### 5.2. Future Direction ###

Some possible avenues for future work:

* Automate projecting the increase in demand for vaccines for those vaccines that requires two shots. In our model, this detail is a bit obfuscated, as we assume the vaccine requirements per age demographic accounts for this. For robustness and improved usability, it would be nice to incorporate this in the data processing sections.

* Introduce slack variables to see how "out-of-the-way" we are going in order to satisfy the vaccine demands. While these requirements are meant to reflect how many people **want** to receive a dose of the vaccine, an easier metric to retreive is how many people **should** get the vaccine (which we assume to be greatly than equal to the requirements outlined in this project).

* Incorporate the concept of wages for the scientists so that the model can more accurately determine when it is appropriate to hire or fire employees.